jmap_mail: load the cache correctly

brong opened the pull request:
Turns out we were calculating how much MORE to cache rather than what to cache up to, but the "loadn" API takes "up to here", not "this much more".  Doh.

wolfsage commented:
I'll leave the code review to someone else - but for the commit message, can we include an explanation of what was wrong and how this fixes it?

Otherwise a year from now we'll just shake our heads at `jmap_mail: load the cache correctly` and wonder

Thanks!

brong commented:
> I'll leave the code review to someone else - but for the commit message, can we include an explanation of what was wrong and how this fixes it?
> 
> Otherwise a year from now we'll just shake our heads at `jmap_mail: load the cache correctly` and wonder
> 
> Thanks!

Yep, done!  I've updated the commit message and also fixed the flipped logic of my last minute rewrite from the one I'd tested by hand on my testbed without good variable names.

rsto commented on imap/jmap_mail.c:
```diff
@@ -3858,24 +3858,29 @@ static void emailquery_cache_slice(struct emailquery *q,
+            // is the window already within the loaded section?
+            if (have_len < need) {
+                // fill just to the end of the window
+                emailquery_cache_loadn(q, qc, need);
```
The `loadn` function is a misnomer I introduced, and now I can now how it is misleading. A better name might be `ensuren` or even `ensure`, like we do in the buf API.

brong replied:
I'll rename it!

rsto reviewed (approved):
Looks good to me, although we might want to rename `loadn`.
